### DISCUSSION TITLE
Possible to get more FiniteMachine instance context in transition logs ?

### DISCUSSION THREAD

**Discussion by raw-bin** [piotrmurach/finite_machine#72] [Q&A]
Hi.

Firstly - thanks for a great gem! :)

I'm using it to model a fairly complex system containing multiple inter-operating entities each with their own state machines that transition in response to message events.

My goals are to use this modelling exercise to:

- Reason about any invalid/redundant state scenarios (the cerebral process of modelling in itself helps me get a better grasp of which entities need to be considered as a part of the system and which don't etc)

- Use the state transition logs + additional logging as the basis for auto-generating API specifications for functions that the entities implement (quite literally a parse, infer and generate type thing)

What I seem to be missing is context in the transition logs about which FiniteMachine instance's states are being transitioned through. As an example, if I create 2 FiniteMachine instances and arrange for suitable events to be triggered for both of them, what emerges in the transition logs is 'trace' that doesn't identify which FiniteMachine instance is involved.

Eg:

```
I, [2021-08-05T11:46:39.661479 #10463]  INFO -- : Transition: @event=init none -> reset
I, [2021-08-05T11:46:39.663105 #10463]  INFO -- : Transition: @event=init none -> reset
I, [2021-08-05T11:46:39.663286 #10463]  INFO -- : Transition: @event=key_on reset -> firmware_running
```

The first 2 events are in the context of different FiniteMachine instances and what I seek is an identifier to also be emitted in the logs.

I note that there is an unused namespace variable here:

https://github.com/piotrmurach/finite_machine/blob/8b1e421e02948f60d10e850c55c585e3cd3ce91a/lib/finite_machine/state_machine.rb#L35

Perhaps that could be the basis for what I need ?

My problem is that I am a Ruby n00b so wasn't sure what would be the best way to do this.

Would be grateful for any advice.

Thanks!

**Comment by piotrmurach** [comment]
Hi Robin 👋 

Thanks for using `finite_machine`. 

The `:namespace` is there for a feature that I didn't have time to implement. I wanted to allow you to scope events inside a namespace so that you could share some event names or even use event names that could potentially conflict with Ruby built-in methods like `fail`. So let's keep it as is.

To implement this feature, I'd suggest adding `:name` keyword to configure and distinguish between instances.  This name then would be passed down to the logger `report_transition` to include it when present. When absent, generate UUID by default with Ruby's [securerandom](https://ruby-doc.org/stdlib-2.5.1/libdoc/securerandom/rdoc/SecureRandom.html).

Are you willing to implement this feature?

**Reply by raw-bin** [reply]
Hi @piotrmurach.

Thanks for the response!

Ok, I'll have a stab at this and report back. Thanks!

**Reply by raw-bin** [reply]
Done! See: https://github.com/piotrmurach/finite_machine/pull/73

Please let me know if you would like me to make any changes.

Thanks.

**Reply by raw-bin** [reply]
Hi @piotrmurach.

I force pushed an update to the same PR with what I think are mods that resolve all your points. Please let me know what you think.

Also: Please note that I have not addressed houndCI's line length complaints. Happy to do that if you would like me to. Just please let me know either way.

Thanks.
